axelor-production/src/main/java/com/axelor/apps/production/service/batch/BatchFillSopRealValues.java

```diff
@@ -0,0 +1,180 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.axelor.apps.production.service.batch;
+
+import com.axelor.apps.base.db.Company;
+import com.axelor.apps.base.db.Period;
+import com.axelor.apps.base.db.ProductCategory;
+import com.axelor.apps.base.db.repo.BatchRepository;
+import com.axelor.apps.base.db.repo.ExceptionOriginRepository;
+import com.axelor.apps.base.exceptions.BaseExceptionMessage;
+import com.axelor.apps.base.service.PeriodService;
+import com.axelor.apps.base.service.administration.AbstractBatch;
+import com.axelor.apps.base.service.app.AppBaseService;
+import com.axelor.apps.base.service.exception.TraceBackService;
+import com.axelor.apps.production.db.ManufOrder;
+import com.axelor.apps.production.db.ProductionBatch;
+import com.axelor.apps.production.db.Sop;
+import com.axelor.apps.production.db.SopLine;
+import com.axelor.apps.production.db.repo.ManufOrderRepository;
+import com.axelor.apps.production.exceptions.ProductionExceptionMessage;
+import com.axelor.apps.sale.db.SaleOrderLine;
+import com.axelor.apps.sale.db.repo.SaleOrderLineRepository;
+import com.axelor.apps.stock.db.repo.StockLocationLineHistoryRepository;
+import com.axelor.i18n.I18n;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.math.RoundingMode;
+
+public class BatchFillSopRealValues extends AbstractBatch {
+
+  protected SaleOrderLineRepository saleOrderLineRepo;
+  protected ManufOrderRepository manufOrderRepo;
+  protected StockLocationLineHistoryRepository stockLocationLineHistoryRepo;
+  protected PeriodService periodService;
+  protected AppBaseService appBaseService;
+
+  @Inject
+  public BatchFillSopRealValues(
+      SaleOrderLineRepository saleOrderLineRepo,
+      ManufOrderRepository manufOrderRepo,
+      StockLocationLineHistoryRepository stockLocationLineHistoryRepo,
+      PeriodService periodService,
+      AppBaseService appBaseService) {
+    this.saleOrderLineRepo = saleOrderLineRepo;
+    this.manufOrderRepo = manufOrderRepo;
+    this.stockLocationLineHistoryRepo = stockLocationLineHistoryRepo;
+    this.periodService = periodService;
+    this.appBaseService = appBaseService;
+  }
+
+  @Override
+  protected void process() {
+    ProductionBatch productionBatch = batch.getProductionBatch();
+    computeSopRealValues(productionBatch);
+  }
+
+  @Transactional
+  protected void computeSopRealValues(ProductionBatch productionBatch) {
+    Company company = productionBatch.getCompany();
+    for (Sop sop : productionBatch.getSopSet()) {
+      computeValuesPerSOP(sop, company);
```

Comment:
Please clear Jpa after every computing or every 10 or 20 computation, will need to do adaptions in somes method